server/src/main/java/org/opensearch/extensions/rest/RestSendToExtensionAction.java

```diff
@@ -121,7 +124,17 @@ public void handleResponse(RestExecuteOnExtensionResponse response) {
                 restExecuteOnExtensionResponse.setStatus(response.getStatus());
                 restExecuteOnExtensionResponse.setContentType(response.getContentType());
                 restExecuteOnExtensionResponse.setContent(response.getContent());
-                restExecuteOnExtensionResponse.setHeaders(response.getHeaders());
+                // Extract the consumed parameters from the header
+                Map<String, List<String>> headers = response.getHeaders();
+                List<String> consumedParams = headers.get(CONSUMED_PARAMS_KEY);
+                if (consumedParams != null) {
+                    consumedParams.stream().forEach(p -> request.param(p));
+                }
+                Map<String, List<String>> headersWithoutConsumedParams = headers.entrySet()
+                    .stream()
+                    .filter(e -> !e.getKey().equals(CONSUMED_PARAMS_KEY))
+                    .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
+                restExecuteOnExtensionResponse.setHeaders(headersWithoutConsumedParams);
```

Comment:
With this design, it seems like the REST request is serialized and transmitted over to the plugin including sensitive headers such as Authorization or Cookie.  Can this be altered so the headers list is recreated with only explicitly allowed fields included?

This would make the parameter consumption logic in the extension SDK straightforward where it doesn't have to consider special cases where something has already been consumed.

Reply:
These are the [custom headers in `RestResponse`](https://github.com/opensearch-project/OpenSearch/blob/15161a86491569d94642f3a88174e1389218b303/server/src/main/java/org/opensearch/rest/RestResponse.java#L50-L52), not the request.

They are usually empty as the only subclass in OpenSearch that uses these is `BytesRestResponse`, which doesn't use them.  They must be explicitly added.  Currently these additions are:
 - [Responding to an `OPTIONS` request](https://github.com/opensearch-project/OpenSearch/blob/15161a86491569d94642f3a88174e1389218b303/server/src/main/java/org/opensearch/rest/RestController.java#L478) or [Unsupported Method error](https://github.com/opensearch-project/OpenSearch/blob/15161a86491569d94642f3a88174e1389218b303/server/src/main/java/org/opensearch/rest/RestController.java#L457) with an Allow header (done by the `RestController` before it ever gets to an extension.)
 - [Adding a Location header from a content builder](https://github.com/opensearch-project/OpenSearch/blob/15161a86491569d94642f3a88174e1389218b303/server/src/main/java/org/opensearch/rest/action/RestStatusToXContentListener.java#L77) in response to a `CREATE` request.
 - [Copying headers from an `OpenSearchException`](https://github.com/opensearch-project/OpenSearch/blob/15161a86491569d94642f3a88174e1389218b303/server/src/main/java/org/opensearch/rest/BytesRestResponse.java#L135) when it is passed to the `BytesRestResponse` constructor.

These headers are already being passed as implemented in a previous PR, the only change here is that I am adding an additional key and extracting it.

Reply:
Thanks for the response, this code looks good :)

Reply:
I've created https://github.com/opensearch-project/OpenSearch/pull/4430 to warn adding disclosing this header information to extensions